lib/rubocop/ast/processed_source.rb

```diff
@@ -240,6 +254,31 @@ def create_parser(ruby_version)
           end
         end
       end
+
+      def first_token_index(range_or_node)
+        begin_pos = source_range(range_or_node).begin_pos
+        sorted_tokens.bsearch_index { |token| token.begin_pos >= begin_pos }
+      end
+
+      def last_token_index(range_or_node)
+        end_pos = source_range(range_or_node).end_pos
+        sorted_tokens.bsearch_index { |token| token.end_pos >= end_pos }
+      end
+
+      # The tokens list is always sorted by token position, except for cases when heredoc
+      # is passed as a method argument. In this case tokens are interleaved by
+      # heredoc contents' tokens.
+      def sorted_tokens
+        @sorted_tokens ||= tokens.sort_by(&:begin_pos)
+      end
+
+      def source_range(range_or_node)
+        if range_or_node.is_a?(RuboCop::AST::Node)
```

Comment:
I thought that `Comment` responded to `:source_range` but it doesn't. I'll open an issue with `parser` coz it should I think :-)
Was there a reason to prefer checking for a class instead of duck typing with `respond_to? :source_range` as I had proposed?

Reply:
ok, made it a duck 😄